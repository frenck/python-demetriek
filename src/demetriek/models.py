"""Models for LaMetric."""
from __future__ import annotations

from datetime import datetime
from ipaddress import IPv4Address
from typing import Any, Optional, Union

from awesomeversion import AwesomeVersion
from pydantic import BaseModel, Field, root_validator

from .const import (
    AlarmSound,
    BrightnessMode,
    DeviceMode,
    DeviceState,
    DisplayType,
    NotificationIconType,
    NotificationPriority,
    NotificationSound,
    NotificationSoundCategory,
    NotificationType,
    WifiMode,
)


class Range(BaseModel):
    """Object holding an integer range."""

    range_min: int = Field(..., alias="min")
    range_max: int = Field(..., alias="max")


class Audio(BaseModel):
    """Object holding the audio state of an LaMetric device."""

    volume: int
    volume_range: Optional[Range]
    volume_limit: Optional[Range]


class Bluetooth(BaseModel):
    """Object holding the Bluetooth state of an LaMetric device."""

    available: bool
    name: str
    active: bool
    discoverable: bool
    pairable: bool
    address: str


class Display(BaseModel):
    """Object holding the display state of an LaMetric device."""

    brightness: int
    brightness_mode: BrightnessMode
    width: int
    height: int
    display_type: DisplayType = Field(None, alias="type")


class Wifi(BaseModel):
    """Object holding the Wi-Fi state of an LaMetric device."""

    active: bool
    mac: str
    available: bool
    encryption: Optional[str] = None
    ssid: str
    ip: IPv4Address
    mode: WifiMode
    netmask: str
    rssi: Optional[int] = None


class Device(BaseModel):
    """Object holding the state of an LaMetric device."""

    device_id: str = Field(..., alias="id")
    name: str
    serial_number: str
    os_version: AwesomeVersion
    mode: DeviceMode
    model: str
    audio: Audio
    bluetooth: Bluetooth
    display: Display
    wifi: Wifi


class Chart(BaseModel):
    """Object holding the chart frame of an LaMetric notification."""

    data: list[int] = Field(..., alias="chartData")

    class Config:
        """Chart model configuration."""

        allow_population_by_field_name = True


class Simple(BaseModel):
    """Object holding the simple frame of an LaMetric notification."""

    icon: Optional[Union[int, str]] = None
    text: str


class GoalData(BaseModel):
    """Object holding the goal data of an LaMetric notification."""

    current: int
    end: int
    start: int
    unit: Optional[str] = None


class Goal(BaseModel):
    """Object holding the goal frame of an LaMetric notification."""

    icon: Optional[Union[int, str]] = None
    data: GoalData = Field(..., alias="goalData")

    class Config:
        """Goal model configuration."""

        allow_population_by_field_name = True


class Sound(BaseModel):
    """Object holding the notification sound state of an LaMetric device."""

    category: Optional[NotificationSoundCategory]
    sound: Union[AlarmSound, NotificationSound] = Field(..., alias="id")
    repeat: int = 1

    @root_validator
    @classmethod
    def infer_category(cls, values: dict[str, Any]) -> dict[str, Any]:  # noqa: F841
        """Infer the category of the sound.

        Args:
            values: The values of the model.

        Returns:
            The values of the model, with the category field inferred.
        """
        if values["category"] is not None:
            return values

        if values["sound"] in AlarmSound:
            values["category"] = NotificationSoundCategory.ALARMS

        if values["sound"] in NotificationSound:
            values["category"] = NotificationSoundCategory.NOTIFICATIONS

        return values

    class Config:
        """Sound model configuration."""

        allow_population_by_field_name = True


class Model(BaseModel):
    """Object holding the notification model of an LaMetric device."""

    frames: list[Union[Chart, Goal, Simple]]
    sound: Optional[Sound] = None
    cycles: int = 1


class Notification(BaseModel):
    """Object holding a LaMetric notification."""

    model: Model
    created: Optional[datetime] = None
    expiration_date: Optional[datetime] = None
    icon_type: Optional[NotificationIconType] = None
    life_time: Optional[float] = None
    priority: Optional[NotificationPriority] = None
    notification_id: Optional[int] = Field(None, alias="id")
    notification_type: Optional[NotificationType] = Field(None, alias="type")


class User(BaseModel):
    """Object holding LaMetric User information."""

    apps_count: int
    email: str
    name: str
    private_apps_count: int
    private_device_count: int
    user_id: int = Field(..., alias="id")


class CloudDevice(BaseModel):
    """Object holding the state of an LaMetric device from the Cloud."""

    device_id: int = Field(..., alias="id")
    name: str
    state: DeviceState
    serial_number: str
    api_key: str
    ip: IPv4Address = Field(..., alias="ipv4_internal")
    mac: str
    ssid: str = Field(..., alias="wifi_ssid")
    created_at: datetime
    updated_at: datetime
