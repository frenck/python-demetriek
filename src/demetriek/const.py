"""Asynchronous Python client for LaMetric TIME devices."""

from enum import Enum


class BrightnessMode(str, Enum):
    """Enum holding the available brightness modes."""

    AUTO = "auto"
    MANUAL = "manual"


class DeviceMode(str, Enum):
    """Enum holding the available device modes."""

    AUTO = "auto"
    KIOSK = "kiosk"
    MANUAL = "manual"
    SCHEDULE = "schedule"


class DeviceState(str, Enum):
    """Enum holding the available device states."""

    BANNED = "banned"
    CONFIGURED = "configured"
    NEW = "new"


class DisplayType(str, Enum):
    """Enum holding the available display types."""

    COLOR = "color"
    GRAYSCALE = "grayscale"
    MIXED = "mixed"
    MONOCHROME = "monochrome"


class NotificationIconType(str, Enum):
    """Enum holding the available icon types."""

    ALERT = "alert"
    INFO = "info"
    NONE = "none"


class NotificationPriority(str, Enum):
    """Enum holding the available notification priorities."""

    CRITICAL = "critical"
    INFO = "info"
    WARNING = "warning"


class NotificationSoundCategory(str, Enum):
    """Enum holding the available notification sound categories."""

    ALARMS = "alarms"
    NOTIFICATIONS = "notifications"


class AlarmSound(str, Enum):
    """Enum holding the available alarm sounds."""

    ALARM1 = "alarm1"
    ALARM2 = "alarm2"
    ALARM3 = "alarm3"
    ALARM4 = "alarm4"
    ALARM5 = "alarm5"
    ALARM6 = "alarm6"
    ALARM7 = "alarm7"
    ALARM8 = "alarm8"
    ALARM9 = "alarm9"
    ALARM10 = "alarm10"
    ALARM11 = "alarm11"
    ALARM12 = "alarm12"
    ALARM13 = "alarm13"


class NotificationSound(str, Enum):
    """Enum holding the available notification sounds."""

    BICYCLE = "bicycle"
    CAR = "car"
    CASH = "cash"
    CAT = "cat"
    DOG = "dog"
    DOG2 = "dog2"
    ENERGY = "energy"
    KNOCK_KNOCK = "knock-knock"
    LETTER_EMAIL = "letter_email"
    LOSE1 = "lose1"
    LOSE2 = "lose2"
    NETGATIVE1 = "negative1"
    NETGATIVE2 = "negative2"
    NETGATIVE3 = "negative3"
    NETGATIVE4 = "negative4"
    NETGATIVE5 = "negative5"
    NOTIFICATION = "notification"
    NOTIFICATION2 = "notification2"
    NOTIFICATION3 = "notification3"
    NOTIFICATION4 = "notification4"
    OPEN_DOOR = "open_door"
    POSITIVE1 = "positive1"
    POSITIVE2 = "positive2"
    POSITIVE3 = "positive3"
    POSITIVE4 = "positive4"
    POSITIVE5 = "positive5"
    POSITIVE6 = "positive6"
    STATISTIC = "statistic"
    THUNDER = "thunder"
    WATER1 = "water1"
    WATER2 = "water2"
    WIN = "win"
    WIN2 = "win2"
    WIND = "wind"
    WIND_SHORT = "wind_short"


class NotificationType(str, Enum):
    """Enum holding the available notification types."""

    INTERNAL = "internal"
    EXTERNAL = "external"


class WifiMode(str, Enum):
    """Enum holding the available Wi-Fi modes."""

    DHCP = "dhcp"
    STATIC = "static"
