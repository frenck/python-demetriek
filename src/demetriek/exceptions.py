"""Exceptions for LaMetric."""


class LaMetricError(Exception):
    """Generic LaMetric exception."""


class LaMetricConnectionError(LaMetricError):
    """LaMetric connection exception."""


class LaMetricAuthenticationError(LaMetricError):
    """LaMetric authentication exception."""


class LaMetricConnectionTimeoutError(LaMetricConnectionError):
    """LaMetric connection Timeout exception."""
