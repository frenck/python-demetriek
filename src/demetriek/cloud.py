"""Asynchronous Python client for LaMetric TIME devices."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from typing import Any

import aiohttp
import async_timeout
import backoff
from aiohttp import hdrs
from pydantic import parse_obj_as
from yarl import URL

from .exceptions import (
    LaMetricConnectionError,
    LaMetricConnectionTimeoutError,
    LaMetricError,
)
from .models import CloudDevice, User


@dataclass
class LaMetricCloud:
    """Main class for handling connections with the LaMetric cloud."""

    token: str
    request_timeout: float = 8.0
    session: aiohttp.client.ClientSession | None = None

    _close_session: bool = False

    @backoff.on_exception(
        backoff.expo, LaMetricConnectionError, max_tries=3, logger=None
    )
    async def _request(
        self,
        uri: str = "",
    ) -> Any:
        """Handle a request to the  LaMetric cloud.

        A generic method for sending/handling HTTP requests done gainst
        the LaMetric cloud.

        Args:
            uri: Request URI, for example `/api/v2/users/me`.

        Returns:
            A Python dictionary (JSON decoded) with the response from the
            LaMetric device.

        Raises:
            LaMetricConnectionError: An error occurred while communication with
                the LaMetric device.
            LaMetricConnectionTimeoutError: A timeout occurred while communicating
                with the LaMetric device.
            LaMetricError: Received an unexpected response from the LaMetric device.
        """
        url = URL.build(scheme="https", host="developer.lametric.com", path=uri)

        if self.session is None:
            self.session = aiohttp.ClientSession()
            self._close_session = True

        headers = {
            "Authorization": f"Bearer {self.token}",
            "Accept": "application/json",
        }

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    hdrs.METH_GET,
                    url,
                    headers=headers,
                    raise_for_status=True,
                )

            content_type = response.headers.get("Content-Type", "")
            if "application/json" not in content_type:
                raise LaMetricError(response.status, {"message": await response.text()})
            return await response.json()

        except asyncio.TimeoutError as exception:
            raise LaMetricConnectionTimeoutError(
                "Timeout occurred while connecting to the LaMetric cloud"
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise LaMetricConnectionError(
                "Error occurred while communicating with the LaMetric cloud"
            ) from exception

    async def current_user(self) -> User:
        """Get LaMetric user information.

        Returns:
            A User object, with information about the current user.
        """
        response = await self._request("/api/v2/me")
        return User.parse_obj(response)

    async def devices(self) -> list[CloudDevice]:
        """Get LaMetric devices from the cloud.

        Returns:
            A list of CloudDevices.
        """
        response = await self._request("/api/v2/users/me/devices")
        return parse_obj_as(list[CloudDevice], response)

    async def device(self, device_id: int) -> CloudDevice:
        """Get a LaMetric device from the cloud.

        Args:
            device_id: The ID of the device to get information for.

        Returns:
            A CloudDevice object, with information about the request device.
        """
        response = await self._request(f"/api/v2/users/me/devices/{device_id}")
        return CloudDevice.parse_obj(response)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> LaMetricCloud:
        """Async enter.

        Returns:
            The LaMetricCloud object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
