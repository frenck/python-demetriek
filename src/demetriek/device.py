"""Asynchronous Python client for LaMetric TIME devices."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from typing import Any, cast

import aiohttp
import async_timeout
import backoff
from aiohttp import hdrs
from aiohttp.helpers import BasicAuth
from pydantic import parse_obj_as
from yarl import URL

from .const import BrightnessMode
from .exceptions import (
    LaMetricAuthenticationError,
    LaMetricConnectionError,
    LaMetricConnectionTimeoutError,
    LaMetricError,
)
from .models import Audio, Bluetooth, Device, Display, Notification, Wifi


@dataclass
class LaMetricDevice:
    """Main class for handling connections with the LaMetric device."""

    host: str
    api_key: str
    request_timeout: float = 8.0
    session: aiohttp.client.ClientSession | None = None

    _close_session: bool = False

    @backoff.on_exception(
        backoff.expo, LaMetricConnectionError, max_tries=3, logger=None
    )
    async def _request(
        self,
        uri: str = "",
        method: str = hdrs.METH_GET,
        data: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to a LaMetric device.

        A generic method for sending/handling HTTP requests done gainst
        the LaMetric device.

        Args:
            uri: Request URI, for example `/api/v2/device`.
            method: HTTP method to use for the request.E.g., "GET" or "POST".
            data: Dictionary of data to send to the LaMetric device.

        Returns:
            A Python dictionary (JSON decoded) with the response from the
            LaMetric device.

        Raises:
            LaMetricAuthenticationError: If the API key is invalid.
            LaMetricConnectionError: An error occurred while communication with
                the LaMetric device.
            LaMetricConnectionTimeoutError: A timeout occurred while communicating
                with the LaMetric device.
            LaMetricError: Received an unexpected response from the LaMetric device.
        """
        url = URL.build(scheme="https", host=self.host, port=4343, path=uri)

        if self.session is None:
            self.session = aiohttp.ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    auth=BasicAuth("dev", self.api_key),
                    headers={"Accept": "application/json"},
                    json=data,
                    raise_for_status=True,
                    ssl=False,
                )

            content_type = response.headers.get("Content-Type", "")
            if "application/json" not in content_type:
                raise LaMetricError(response.status, {"message": await response.text()})
            return await response.json()

        except asyncio.TimeoutError as exception:
            raise LaMetricConnectionTimeoutError(
                f"Timeout occurred while connecting to the LaMetric device at {self.host}"
            ) from exception
        except aiohttp.ClientResponseError as exception:
            if exception.status in [401, 403]:
                raise LaMetricAuthenticationError(
                    f"Authentication to the LaMetric device at {self.host} failed"
                ) from exception
            raise LaMetricError(
                f"Error occurred while connecting to the LaMetric device at {self.host}"
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise LaMetricConnectionError(
                f"Error occurred while communicating with the LaMetric device at {self.host}"
            ) from exception

    async def device(self) -> Device:
        """Get LaMetric device information.

        Returns:
            A Device object, with information about the LaMetric device.
        """
        response = await self._request("/api/v2/device")

        response["wifi"].update(
            mac=response["wifi"].get("address"),
            ssid=response["wifi"].get("essid"),
            rssi=response["wifi"].get("strength"),
        )

        return Device.parse_obj(response)

    async def display(
        self,
        *,
        brightness: int | None = None,
        brightness_mode: BrightnessMode | None = None,
    ) -> Display:
        """Get or set LaMetric device display information.

        Args:
            brightness: Brightness level to set.
            brightness_mode: Brightness mode to set.

        Returns:
            A Display object, with latest or updated information about
            the display of the LaMetric device.
        """
        data: dict[str, int | BrightnessMode] = {}

        if brightness is not None:
            data["brightness"] = brightness

        if brightness_mode is not None:
            data["brightness_mode"] = brightness_mode

        if data:
            response = await self._request(
                "/api/v2/device/display", method=hdrs.METH_PUT, data=data
            )
            return Display.parse_obj(response["success"]["data"])

        response = await self._request("/api/v2/device/display")
        return Display.parse_obj(response)

    async def audio(self, *, volume: int | None = None) -> Audio:
        """Get or set LaMetric device audio information.

        Args:
            volume: Volume level to set.

        Returns:
            An Audio object, with latest or updated information about the
            audio state of the LaMetric device.
        """
        data: dict[str, int] = {}

        if volume is not None:
            data["volume"] = volume

        if data:
            response = await self._request(
                "/api/v2/device/audio", method=hdrs.METH_PUT, data=data
            )
            return Audio.parse_obj(response["success"]["data"])

        data = await self._request("/api/v2/device/audio")
        return Audio.parse_obj(data)

    async def bluetooth(self, *, active: bool | None = None) -> Bluetooth:
        """Get LaMetric device bluetooth information.

        Args:
            active: Whether to activate or deactivate Bluetooth.

        Returns:
            A Bluetooth object, with the latest or updated Bluetooth information.
        """
        data: dict[str, int] = {}

        if active is not None:
            data["active"] = active

        if data:
            response = await self._request(
                "/api/v2/device/bluetooth", method=hdrs.METH_PUT, data=data
            )
            response = response["success"]["data"]
        else:
            response = await self._request("/api/v2/device/bluetooth")
        response.update(address=response.get("mac"))
        return Bluetooth.parse_obj(response)

    async def wifi(self) -> Wifi:
        """Get LaMetric device bluetooth information.

        Returns:
            A Wifi object with the latest Wi-Fi state of the device.
        """
        data = await self._request("/api/v2/device/wifi")
        data.update(ip=data.get("ipv4"), rssi=data.get("signal_strength"))
        return Wifi.parse_obj(data)

    async def app_next(self) -> None:
        """Switch to the next app on LaMetric Time.

        App order is controlled by the user via LaMetric Time app.
        """
        await self._request("/api/v2/device/apps/next", method=hdrs.METH_PUT)

    async def app_previous(self) -> None:
        """Switch to the next app on LaMetric Time.

        App order is controlled by the user via LaMetric Time app.
        """
        await self._request("/api/v2/device/apps/prev", method=hdrs.METH_PUT)

    async def notify(
        self,
        *,
        notification: Notification,
    ) -> int:
        """Send a notification to a LaMetric device.

        Args:
            notification: A Notification object.

        Returns:
            The ID of the notification.
        """
        response = await self._request(
            "/api/v2/device/notifications",
            method=hdrs.METH_POST,
            data=notification.dict(
                by_alias=True,
                exclude_none=True,
            ),
        )
        return cast(int, response["success"]["id"])

    async def dismiss_notification(self, *, notification_id: int) -> None:
        """Remove a notification from the queue.

        In case if it is already visible - dismisses it.

        Args:
            notification_id: Notification ID to dismiss.
        """
        await self._request(
            f"/api/v2/device/notifications/{notification_id}",
            method=hdrs.METH_DELETE,
        )

    async def dismiss_all_notifications(self) -> None:
        """Dismiss all notifications notification."""
        if not (notifications := await self.notification_queue()):
            return

        # Dismiss notifications in reverse order to avoid them showing up
        # during rapid dismissal.
        for notification in reversed(notifications):
            if notification.notification_id:
                await self.dismiss_notification(
                    notification_id=notification.notification_id
                )

    async def dismiss_current_notification(self) -> None:
        """Dismiss current notification."""
        if (
            notification := await self.notification_current()
        ) and notification.notification_id:
            await self.dismiss_notification(
                notification_id=notification.notification_id
            )

    async def notification_current(self) -> Notification | None:
        """Get the current notification.

        Returns:
            A Notification objects.
        """
        if data := await self._request("/api/v2/device/notifications/current"):
            return parse_obj_as(Notification, data)
        return None

    async def notification_queue(self) -> list[Notification]:
        """Get the list of all notifications in the queue.

        Notifications with higher priority will be first in the list.

        Returns:
            A list of Notification objects.
        """
        data = await self._request("/api/v2/device/notifications")
        return parse_obj_as(list[Notification], data)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> LaMetricDevice:
        """Async enter.

        Returns:
            The LaMetricDevice object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
