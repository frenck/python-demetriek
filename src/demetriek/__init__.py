"""Asynchronous Python client for LaMetric TIME devices."""
from .cloud import LaMetricCloud
from .const import (
    AlarmSound,
    BrightnessMode,
    DeviceMode,
    DeviceState,
    DisplayType,
    NotificationIconType,
    NotificationPriority,
    NotificationSound,
    NotificationSoundCategory,
    NotificationType,
    WifiMode,
)
from .device import LaMetricDevice
from .exceptions import (
    LaMetricAuthenticationError,
    LaMetricConnectionError,
    LaMetricConnectionTimeoutError,
    LaMetricError,
)
from .models import (
    Audio,
    Bluetooth,
    Chart,
    CloudDevice,
    Device,
    Display,
    Goal,
    GoalData,
    Model,
    Notification,
    Range,
    Simple,
    Sound,
    User,
    Wifi,
)

__all__ = [
    "AlarmSound",
    "Audio",
    "Bluetooth",
    "BrightnessMode",
    "Chart",
    "CloudDevice",
    "Device",
    "DeviceMode",
    "DeviceState",
    "Display",
    "DisplayType",
    "Goal",
    "GoalData",
    "LaMetricCloud",
    "LaMetricAuthenticationError",
    "LaMetricConnectionError",
    "LaMetricConnectionTimeoutError",
    "LaMetricDevice",
    "LaMetricError",
    "Model",
    "Notification",
    "NotificationIconType",
    "NotificationPriority",
    "NotificationSound",
    "NotificationSoundCategory",
    "NotificationType",
    "Range",
    "Simple",
    "Sound",
    "User",
    "Wifi",
    "WifiMode",
]
