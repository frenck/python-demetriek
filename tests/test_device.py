"""Asynchronous Python client for LaMetric TIME devices."""

from dataclasses import asdict

import aiohttp
import pytest
from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from demetriek import (
    Chart,
    Goal,
    GoalData,
    LaMetricDevice,
    Model,
    Notification,
    NotificationIconType,
    NotificationSound,
    Simple,
    Sound,
)
from demetriek.const import (
    NotificationType,
)

from . import load_fixture


@pytest.mark.parametrize(
    "fixture",
    [
        "device.json",
        "device2.json",
        "device3.json",
    ],
)
async def test_get_device(
    aresponses: ResponsesMockServer, fixture: str, snapshot: SnapshotAssertion
) -> None:
    """Test getting device information."""
    aresponses.add(
        "127.0.0.2:4343",
        "/api/v2/device",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture(fixture),
        ),
    )
    async with aiohttp.ClientSession() as session:
        demetriek = LaMetricDevice(host="127.0.0.2", api_key="abc", session=session)
        device = await demetriek.device()

    assert asdict(device) == snapshot


async def test_notify(aresponses: ResponsesMockServer) -> None:
    """Test sending notification serialization."""
    aresponses.add(
        "127.0.0.2:4343",
        "/api/v2/device/notifications",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("notification.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        demetriek = LaMetricDevice(host="127.0.0.2", api_key="abc", session=session)
        notification = Notification(
            icon_type=NotificationIconType.ALERT,
            notification_type=NotificationType.EXTERNAL,
            model=Model(
                frames=[
                    Simple(text="Yeah", icon=18815),
                    Goal(
                        icon=7956,
                        data=GoalData(
                            current=65,
                            end=100,
                            start=0,
                            unit="%",
                        ),
                    ),
                    Chart(data=[1, 2, 3, 4, 5, 4, 3, 2, 1]),
                ],
                sound=Sound(sound=NotificationSound.WIN),
            ),
        )
        response = await demetriek.notify(notification=notification)

    # check response
    assert response == 1
    # check on serialized request if aliases are used and null values are removed
    request = await aresponses.history[0].request.json()
    assert request["type"] == "external"
    assert request["icon_type"] == "alert"
    assert "life_time" not in request
    assert request["model"]["sound"]["id"] == "win"
    assert request["model"]["sound"]["category"] == "notifications"
    assert request["model"]["frames"][0]["text"] == "Yeah"
    assert request["model"]["frames"][1]["goalData"]["current"] == 65
    assert request["model"]["frames"][2]["chartData"] == [1, 2, 3, 4, 5, 4, 3, 2, 1]
