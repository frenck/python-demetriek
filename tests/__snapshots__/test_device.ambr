# serializer version: 1
# name: test_get_device[device.json]
  dict({
    'audio': dict({
      'available': True,
      'volume': 100,
      'volume_limit': dict({
        'range_max': 100,
        'range_min': 0,
      }),
      'volume_range': dict({
        'range_max': 100,
        'range_min': 0,
      }),
    }),
    'bluetooth': dict({
      'active': False,
      'address': 'AA:BB:CC:DD:EE:FF',
      'available': True,
      'discoverable': True,
      'name': 'LM1234',
      'pairable': True,
    }),
    'device_id': '12345',
    'display': dict({
      'brightness': 100,
      'brightness_limit': dict({
        'range_max': 100,
        'range_min': 2,
      }),
      'brightness_mode': <BrightnessMode.AUTO: 'auto'>,
      'brightness_range': dict({
        'range_max': 100,
        'range_min': 0,
      }),
      'display_type': <DisplayType.MIXED: 'mixed'>,
      'height': 8,
      'on': None,
      'screensaver': dict({
        'enabled': False,
      }),
      'width': 37,
    }),
    'mode': <DeviceMode.AUTO: 'auto'>,
    'model': 'LM 37X8',
    'name': "Frenck's LaMetric",
    'os_version': '2.2.2',
    'serial_number': 'SA110405124500W00BS9',
    'update': None,
    'wifi': dict({
      'active': True,
      'available': True,
      'encryption': 'WPA',
      'ip': IPv4Address('192.168.1.11'),
      'mac': 'AA:BB:CC:DD:EE:FF',
      'mode': <WifiMode.DHCP: 'dhcp'>,
      'netmask': '255.255.255.0',
      'rssi': 21,
      'ssid': 'Frenck',
    }),
  })
# ---
# name: test_get_device[device2.json]
  dict({
    'audio': dict({
      'available': True,
      'volume': 100,
      'volume_limit': dict({
        'range_max': 100,
        'range_min': 0,
      }),
      'volume_range': dict({
        'range_max': 100,
        'range_min': 0,
      }),
    }),
    'bluetooth': dict({
      'active': False,
      'address': 'AA:BB:CC:DD:EE:FF',
      'available': True,
      'discoverable': True,
      'name': 'LM1234',
      'pairable': True,
    }),
    'device_id': '12345',
    'display': dict({
      'brightness': 100,
      'brightness_limit': dict({
        'range_max': 100,
        'range_min': 2,
      }),
      'brightness_mode': <BrightnessMode.AUTO: 'auto'>,
      'brightness_range': dict({
        'range_max': 100,
        'range_min': 0,
      }),
      'display_type': <DisplayType.MIXED: 'mixed'>,
      'height': 8,
      'on': None,
      'screensaver': dict({
        'enabled': False,
      }),
      'width': 37,
    }),
    'mode': <DeviceMode.SCHEDULE: 'schedule'>,
    'model': 'LM 37X8',
    'name': "Frenck's LaMetric",
    'os_version': '2.2.2',
    'serial_number': 'SA110405124500W00BS9',
    'update': None,
    'wifi': dict({
      'active': True,
      'available': True,
      'encryption': None,
      'ip': IPv4Address('192.168.1.11'),
      'mac': 'AA:BB:CC:DD:EE:FF',
      'mode': <WifiMode.DHCP: 'dhcp'>,
      'netmask': '255.255.255.0',
      'rssi': None,
      'ssid': 'Frenck',
    }),
  })
# ---
# name: test_get_device[device3.json]
  dict({
    'audio': dict({
      'available': True,
      'volume': 46,
      'volume_limit': dict({
        'range_max': 100,
        'range_min': 0,
      }),
      'volume_range': dict({
        'range_max': 100,
        'range_min': 0,
      }),
    }),
    'bluetooth': dict({
      'active': False,
      'address': '40:F4:C9:AA:AA:AA',
      'available': True,
      'discoverable': True,
      'name': 'LM8367',
      'pairable': False,
    }),
    'device_id': '67790',
    'display': dict({
      'brightness': 100,
      'brightness_limit': dict({
        'range_max': 100,
        'range_min': 2,
      }),
      'brightness_mode': <BrightnessMode.AUTO: 'auto'>,
      'brightness_range': dict({
        'range_max': 100,
        'range_min': 0,
      }),
      'display_type': <DisplayType.MIXED: 'mixed'>,
      'height': 8,
      'on': True,
      'screensaver': dict({
        'enabled': True,
      }),
      'width': 37,
    }),
    'mode': <DeviceMode.MANUAL: 'manual'>,
    'model': 'sa8',
    'name': 'TIME',
    'os_version': '3.1.3',
    'serial_number': 'SASOMETHING',
    'update': dict({
      'version': '3.2.1',
    }),
    'wifi': dict({
      'active': True,
      'available': True,
      'encryption': 'WPA',
      'ip': IPv4Address('192.168.0.33'),
      'mac': '40:F4:C9:AA:AA:AA',
      'mode': <WifiMode.DHCP: 'dhcp'>,
      'netmask': '255.255.255.0',
      'rssi': 78,
      'ssid': 'NETGEARAA',
    }),
  })
# ---
